Improve storyboard ux

manuelmeister opened the pull request:
|before|after|
|-|-|
| ![Bildschirmfoto 2023-01-07 um 22 17 15](https://user-images.githubusercontent.com/3001985/211170657-025141a3-a434-47fb-b15c-632de5f920b8.png)| ![Bildschirmfoto 2023-01-07 um 22 18 31](https://user-images.githubusercontent.com/3001985/211170677-08e5f205-a957-4c94-b921-bf64d2a05ba6.png) |
| ![Bildschirmfoto 2023-01-07 um 22 19 23](https://user-images.githubusercontent.com/3001985/211170695-d8628bde-5ead-4d01-8095-50088ccd405e.png) | ![Bildschirmfoto 2023-01-07 um 22 20 02](https://user-images.githubusercontent.com/3001985/211170715-4065e230-c0cf-4bdc-b857-5eb24912e7c2.png) |

manuelmeister committed:
Improve storyboard ux

BacLuc commented on frontend/src/components/activity/content/StoryboardRow.vue:
```diff
@@ -0,0 +1,188 @@
+}
+
+.e-storyboard-row__delete {
+  color: rgba(0, 0, 0, 0.54) !important;
```
do we need the important here?

manuelmeister replied:
Yes, I'm overwriting the VBtn behaviour (which features important as well)

BacLuc commented on frontend/src/components/activity/content/StoryboardRow.vue:
```diff
@@ -0,0 +1,188 @@
+  color: rgba(0, 0, 0, 0.54) !important;
+
+  &:hover {
+    color: #d32f2f !important;
```
do we need the important here?

manuelmeister replied:
Yes, I'm overwriting the VBtn behaviour (which features important as well)

BacLuc commented on frontend/src/components/activity/content/Storyboard.vue:
```diff
@@ -165,8 +108,17 @@ export default {
       return Object.keys(this.sections).length === 1
     },
+    variant() {
+      return this.clientWidth <= 910 ? 'dense' : 'default'
```
pity that it did not work with css only.
Are media queries (or an abstraction of media query) not the way to go for this problem anymore?

manuelmeister replied:
If someday we have support for container queries, then this could be doable

BacLuc commented on frontend/src/scss/global.scss:
```diff
@@ -98,3 +98,7 @@ body {
 }
+
+.d-contents {
+  display: contents;
```
for what do we need this?

manuelmeister replied:
I need it everytime while prototyping. It is in line with the display helpers of vuetify

BacLuc reviewed (commented):
Putting time and responsible on a separate row improves the look and feel on mobile a lot.
I am not sure if removing the arrows on desktop is the right way, but you will have you reasons.

But on desktop, the drag handle and the delete button is now close.
On the print ui, you put the drag handle on the left side, maybe to avoid that problem.

usu commented on frontend/src/components/form/tiptap/VTiptapEditor.vue:
```diff
@@ -78,6 +78,7 @@ div.v-text-field__slot {
   .editor,
 .v-text-field.v-text-field--enclosed.v-text-field--outlined:not(.v-input--dense) .editor {
-  margin-top: 10px;
+  margin-top: 6px;
```
This leads to a mismatch between cursor and placeholder text. Can we keep the 10px?

<img width="194" alt="image" src="https://user-images.githubusercontent.com/449555/214370903-6964767c-d447-422c-97fa-ab75cfee9e26.png">

manuelmeister replied:
Fixed

usu commented on frontend/src/components/activity/content/Storyboard.vue:
```diff
@@ -1,124 +1,65 @@
-        <v-col cols="1" />
-      </v-row>
-
-      <api-sortable
```
ApiSortable is not used anywhere else. Couldn't we just adjust ApiSortable with whatever functionality was missing, instead of duplicating to StoryboardSortable?

manuelmeister replied:
Sadly no, if we want to keep api sortable generic. Storyboard sortable needs to be the correct structure for the screenreader, otherwise it won't make sense for people using a screen reader.

If you want I can delete ApiSortable.

usu commented on frontend/src/components/activity/content/Storyboard.vue:
```diff
@@ -1,124 +1,65 @@
+  <card-content-node v-resizeobserver.debounce="onResize" v-bind="$props">
+    <component :is="variant === 'default' ? 'table' : 'div'">
+      <thead v-if="variant === 'default'">
+        <tr>
```
We really want this to be a normal table and not v-row/v-col components as before?

manuelmeister replied:
Pretty sure. Otherwise we need to reimplement a11y for the whole thing again.

usu replied:
I'm unfortunately not really knowledgeable an a11y, so cannot really judge on this. What seems strange to me is that we use table elements for something that doesn't really look like a table (in dense variant) and where vuetify provided the v-row/v-col for exactly this purpose. 

Also in default variant there's an extra div in the table:
![image](https://user-images.githubusercontent.com/449555/223143627-f54cc715-9ab4-4f63-9d3c-238565674da7.png)

And in dense variant there's even no table element:
![image](https://user-images.githubusercontent.com/449555/223143832-66208baa-495c-4107-b678-74837a5348ee.png)

As far as I know, both are not valid HTML.

manuelmeister replied:
This should be sorted now

usu reviewed (commented):
Looks nice, visually! And agree to the comments from @BacLuc

manuelmeister commented:
## Core Meeting Decision

- Drag handles all the way
- Up and down only on focus

manuelmeister committed:
Merge branch 'devel' into improve/storyboard

manuelmeister committed:
Merge remote-tracking branch 'upstream/devel' into improve/storyboard

# Conflicts:
#	frontend/package-lock.json
#	frontend/package.json

manuelmeister committed:
Simplify reordering of storyboard entries

manuelmeister commented:
@BacLuc & @usu can you look at it again? I've fixed the requested changes.

<img width="868" alt="Bildschirmfoto 2023-03-06 um 19 21 10" src="https://user-images.githubusercontent.com/3001985/223108949-9f229ebf-386c-4677-be92-6699a290bebf.png">
<img width="1031" alt="Bildschirmfoto 2023-03-06 um 19 21 35" src="https://user-images.githubusercontent.com/3001985/223109024-052780c1-0f90-4a1a-971f-08ea1ac37b93.png">

usu reviewed (commented):
Just checked visually, haven't looked into the code yet. Visual appearance looks good to me. Points below are the ones I noticed so far.

- Without any content, all the columns are smashed together.
![image](https://user-images.githubusercontent.com/449555/223130634-1512a044-fd18-456e-9c41-c6e44fcef396.png)

- In dense variant, the tab order is still time -> program -> responsible, which is visually unexpected

- Arrows are completely gone. By purpose? Cannot fully recall the discussion in January, but the notes say "Up and down only on focus".

carlobeltrame commented:
There is also this sentry issue which popped up on the feature branch deployment: https://ecamp.sentry.io/issues/3980035515/
Someone on Windows had this issue 3 times around 2pm / 3pm. No commits have been added since then, so I assume this is still relevant and should be fixed?

manuelmeister commented:
@usu

> Without any content, all the columns are smashed together.

I thought we had configured that it always needs at least one row.

> In dense variant, the tab order is still time -> program -> responsible, which is visually unexpected

I agree, but taborder should not be changed. Also it is not easily changeable via vuetify. I added a if statement to combat this.

> Arrows are completely gone. By purpose? Cannot fully recall the discussion in January, but the notes say "Up and down only on focus".

Yes, I changed it to be movable with the arrow keys while on the move handle. This way, it is a power user feature and does not disturb the tabflow anymore than the move handle and the delete button already do. I tried with an overlay but it required far more work. I could put in in a new issue, if the new way is not the preferred way.

manuelmeister committed:
Fix semantic and order

usu commented:
> > Without any content, all the columns are smashed together.
> 
> I thought we had configured that it always needs at least one row.

Ah yes, true. In this case more of an edge case with our dev data.

usu commented on frontend/src/components/activity/content/StoryboardRow.vue:
```diff
@@ -0,0 +1,210 @@
+<template>
+  <component
+    :is="variant !== 'dense' ? 'tr' : 'div'"
```
Ui, this component has now really grown in size and complexity. I see what you tried to achieve. But with all the `<component>` tags and `v-if` and ternary operators it's very difficult to grasp on what's happening in this component.

I'd definitely prefer "composition over configuration" in this instance. Visually, it look's good for me though. I'm also ok to merge and I can propose a refactor afterwards. Or I can make a proposal as PR to this PR. As you prefer.

manuelmeister replied:
I agree as well. A11y components often are more complex. I assume you would want to extract the dense version?

I think this could be done in a improvement task. I believe, this UI improvement should land on stage for the closed beta, as it greatly improves the mobile view.

manuelmeister replied:
Done

usu commented on frontend/src/components/activity/content/StoryboardRow.vue:
```diff
@@ -0,0 +1,210 @@
+        class="drag-and-drop-handle"
+        :disabled="isLastSection"
+        :aria-label="$tc('components.activity.content.storyboardRow.move')"
+        @keydown.down="$emit('moveDown', itemKey)"
```
Ah I see, here are the keyup/keydown. Definitely a "power-user" thing and didn't work well for me even after knowing the feature. I thought we wanted to keep the arrows especially for mobile, where this hack doesn't work. But also for this, I'm ok to merge and we can improve/discuss later.

manuelmeister replied:
I agree. I think this power feature would need some work anyway, because if you move an entry down, the focus should stay on the handle. We could also show a hint when the handle ist focused that you can move the entry with the arrow keys.

We tried tried drag&drop on different mobile platforms and were surprised that it worked so well. So we decided to use the arrows only on focus. But in my experiments I made an overlay over the whole row that was only visible when focused and this felt strange and cumbersome, as it introduced two tabstops per row that are initially hidden. I think we can find out in the user feedback.

usu commented on common/locales/de.json:
```diff
@@ -62,6 +62,7 @@
     "storyboard": {
       "entity": {
         "section": {
+          "controls": "Aktionen",
```
I don't think this should be in common. Common is meant for shared keys between print and frontend and I don't see us using this translation key in print. Also "entity" subkeys were so far strictly reserved for field names and entity names.

Could be in `components.activity.content.storyboard.controls` (same location as `reorder`). Also fine for me to have this in `globals.controls`, as we probably will use the exact same key in many more locations, if we apply `sr-only` more consistently.

manuelmeister committed:
Move controls key to frontend

manuelmeister committed:
Extract default and dense storyboard row variant

manuelmeister committed:
Fix i18n lint

manuelmeister commented:
@carlobeltrame the issue you've described is a performance problem and probably needs to be looked at. However it is not critical as it just doesn't update to the other variant.
https://stackoverflow.com/questions/49384120/resizeobserver-loop-limit-exceeded

carlobeltrame commented on frontend/src/locales/de.json:
```diff
@@ -8,6 +8,10 @@
         },
+        "storyboard": {
+          "controls": "Aktionen",
+          "reorder": "Umsortieren"
```
This is only for screen readers, right?
Otherwise I'd propose to call this "Reihenfolge ändern". But if it is supposed to describe only a single currently focused drag handle, this should be fine (or maybe "Verschieben" could be more standard German).

manuelmeister replied:
```suggestion
          "reorder": "Verschieben"
```

carlobeltrame commented on frontend/src/locales/fr.json:
```diff
@@ -8,6 +8,10 @@
         },
+        "storyboard": {
+          "controls": "Actions",
+          "reorder": "Reorder"
```
Is this in English on purpose?

manuelmeister replied:
```suggestion
          "reorder": "Déplacer"
```

carlobeltrame reviewed (approved):
Love it! I really like how it even looks great on desktop if I have a narrow storyboard (in a narrow column of a columnLayout).

The reordering and maybe deleting could use a save progress indicator some time in the future.

As a guest in the camp, I still see the drag handles and delete buttons, but when I try to use them I get Access Denied toasts. Similarly, If I try to edit the storyboard sections, our red retry and cancel buttons appear. This is fine for now, but maybe we can disable the controls or simplify the view for guests in the future.

manuelmeister commented on frontend/src/locales/it.json:
```diff
@@ -8,6 +8,10 @@
         },
+        "storyboard": {
+          "controls": "Azioni",
+          "reorder": "Riordino"
```
```suggestion
          "reorder": "Spostare"
```

manuelmeister committed:
Change to "move" in other languages

manuelmeister committed:
Merge branch 'devel' into improve/storyboard
